### ISSUE TITLE
Implement windows support

### ISSUE THREAD

**Issue by dbuenzli** [dbuenzli/mtime#2]
Using [performance counters](https://msdn.microsoft.com/en-us/library/windows/desktop/aa373083%28v=vs.85%29.aspx). PR welcome.

**Comment by whitequark** [comment]
Not a PR but a method that does exactly what you want: https://github.com/whitequark/solvespace/blob/master/src/win32/w32main.cpp#L661-L668

(I would send a PR but I have no idea how to build OCaml on my Windows machine.)

**Comment by dbuenzli** [comment]
Thanks for the reference. All this should be easy enough to do in a VM with `ocpwin` but I lack time right now, this summer maybe.

**Comment by hcarty** [comment]
This is a patch I'm using locally.  I'm happy to create a PR if you're interested.

Some caveats:

- ms granularity due to the API call used.  ns granularity is not important for the project I initially created this patch for
- I've only tested it using opam-cross-windows + MXE to compile, running the result on Windows 7 64bit VMs
- I have no idea if the subtraction of `1` from `start` is necessary

```patch
diff --git a/src-os/mtime_stubs.c b/src-os/mtime_stubs.c
index 35c1406..d79124f 100644
--- a/src-os/mtime_stubs.c
+++ b/src-os/mtime_stubs.c
@@ -10,7 +10,10 @@
 
 /* Detect platform */
 
-#if defined(__APPLE__) && defined(__MACH__)
+#if defined(_WIN32)
+ #define OCAML_MTIME_WIN32
+
+#elif defined(__APPLE__) && defined(__MACH__)
   #define OCAML_MTIME_DARWIN
 
 #elif defined (__unix__) || defined(__unix)
@@ -63,6 +66,27 @@ CAMLprim value ocaml_mtime_elapsed_ns (value unit)
                           (uint64_t)(now.tv_nsec - start.tv_nsec));
 }
 
+/* Windows */
+
+#elif defined(OCAML_MTIME_WIN32)
+
+#include <windows.h>
+
+CAMLprim value ocaml_mtime_elapsed_ns (value unit)
+{
+  static ULONGLONG start = 0L;
+  if (start == 0L)
+  {
+    /* XXX: This provides ms granularity at best.
+     * Subtracting one to guarantee an initial value
+     * other than 0. */
+    start = GetTickCount64 () - 1L;
+  }
+
+  ULONGLONG now = GetTickCount64 ();
+  return caml_copy_int64 ((now - start) * (ULONGLONG)1000);
+}
+
 /* Unsupported */
 
 #else
```

**Comment by dbuenzli** [comment]
Here's another patch by @fdopen https://github.com/fdopen/opam-repository-mingw/blob/master/packages/mtime/mtime.0.8.3/files/mtime-0.8.3.patch

I don't know why windows people don't upstream patches....

**Comment by hcarty** [comment]
> I don't know why windows people don't upstream patches....

To be fair, my comment was an offer to do exactly that.  But if it works then @fdopen's patch looks like a better approach.

**Comment by dbuenzli** [comment]
It seems the patch has been updated for the latest version https://github.com/fdopen/opam-repository-mingw/commit/4c6537c0138f493ddf705057cb917eecf77e64f1

**Comment by dbuenzli** [comment]
Fixed by #35
